src/mscorlib/src/System/CurrentSystemTimeZone.cs

```diff
@@ -0,0 +1,354 @@
+                return TimeSpan.Zero;
+            }
+            else
+            {
+                return new TimeSpan(TimeZone.CalculateUtcOffset(time, GetDaylightChanges(time.Year)).Ticks + m_ticksOffset);
+            }
+        }
+
+        //
+        // Return the (numberOfSunday)th day of week in a particular year/month.
+        //
+        private static DateTime GetDayOfWeek(int year, bool fixedDate, int month, int targetDayOfWeek, int numberOfSunday, int hour, int minute, int second, int millisecond)
+        {
+            DateTime time;
+
+            if (fixedDate)
+            {
+                //
+                // Create a Fixed-Date transition time based on the supplied parameters
+                // For Fixed-Dated transition times, the 'numberOfSunday' parameter actually
+                // represents the day of the month.
+                //
+
+                // if the day is out of range for the month then use the last day of the month
+                int day = DateTime.DaysInMonth(year, month);
+
+                time = new DateTime(year, month, (day < numberOfSunday) ? day : numberOfSunday,
+                                  hour, minute, second, millisecond, DateTimeKind.Local);
+            }
+            else if (numberOfSunday <= 4)
+            {
+                //
+                // Get the (numberOfSunday)th Sunday.
+                //
+
+                time = new DateTime(year, month, 1, hour, minute, second, millisecond, DateTimeKind.Local);
+
+                int dayOfWeek = (int)time.DayOfWeek;
+                int delta = targetDayOfWeek - dayOfWeek;
+                if (delta < 0)
+                {
+                    delta += 7;
+                }
+                delta += 7 * (numberOfSunday - 1);
+
+                if (delta > 0)
+                {
+                    time = time.AddDays(delta);
+                }
+            }
+            else
+            {
+                //
+                // If numberOfSunday is greater than 4, we will get the last sunday.
+                //
+                Calendar cal = GregorianCalendar.GetDefaultInstance();
+                time = new DateTime(year, month, cal.GetDaysInMonth(year, month), hour, minute, second, millisecond, DateTimeKind.Local);
+                // This is the day of week for the last day of the month.
+                int dayOfWeek = (int)time.DayOfWeek;
+                int delta = dayOfWeek - targetDayOfWeek;
+                if (delta < 0)
+                {
+                    delta += 7;
+                }
+
+                if (delta > 0)
+                {
+                    time = time.AddDays(-delta);
+                }
+            }
+            return (time);
+        }
+
+        [System.Security.SecurityCritical]  // auto-generated
+        [ResourceExposure(ResourceScope.None)]
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal extern static int nativeGetTimeZoneMinuteOffset();
```

Comment:
@ellismg I already asked @hann013 regarding Linux based implementation and he mentioned will pick Mono implementation.